src/main/java/br/com/sunty/controller/course/CourseAdminController.java

```diff
@@ -1,55 +1,123 @@
+    public CourseAdminController(CourseRepository courseRepository, SubCategoryRepository subCategoryRepository, InstructorRepository instructorRepository) {
         this.courseRepository = courseRepository;
         this.subCategoryRepository = subCategoryRepository;
+        this.instructorRepository = instructorRepository;
     }
 
 
     @GetMapping("/admin/courses/{categoryUrlCode}/{subcategoryUrlCode}")
     public String findAll(@PathVariable String subcategoryUrlCode,
+                          @PathVariable String categoryUrlCode,
                           @PageableDefault(size = 5)
                           Pageable pageable,
                           Model model) {
 
-        SubCategory subCategory = subCategoryRepository.findByUrlCode(subcategoryUrlCode)
-                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, subcategoryUrlCode));
+        AdminSubCategoryDto adminSubCategoryDto = subCategoryRepository.findByUrlCode(subcategoryUrlCode)
+                .map(AdminSubCategoryDto::new)
+                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, subcategoryUrlCode));
+
+        SubCategory subcategory = subCategoryRepository.findByUrlCode(subcategoryUrlCode)
+                .orElseThrow(() -> new ResponseStatusException(NOT_FOUND, subcategoryUrlCode));
 
-        Page<Course> courses = courseRepository.findAllBySubCategory(subCategory, pageable);
+        Page<Course> courses = courseRepository.findAllBySubCategory(subcategory, pageable);
         Page<AdminCourseDto> courseDtos = courses.map(AdminCourseDto::new);
 
-        model.addAttribute("subCategory", subCategory.getName());
+        model.addAttribute("subCategory", adminSubCategoryDto);
         model.addAttribute("courses", courseDtos);
         return "course/courseList";
     }
 
-    @GetMapping("/admin/courses/{categoryUrlCode}/{subcategoryUrlCode}/new")
-    public String redirectNew(){
+    @GetMapping("/admin/courses/new")
+    public String createForm(Model model) {
+        List<Instructor> instructors = instructorRepository.findAll();
+        List<SubCategory> subCategories = subCategoryRepository.findAll();
+
+        model.addAttribute("subCategories", subCategories);
+        model.addAttribute("instructors", instructors);
         return "course/newCourseForm";
     }
 
-    @GetMapping("/admin/courses/{courseurlCode}")
-    public String redirectEdit(){
+    @PostMapping("/admin/courses")
+    public String create(@Valid AdminNewCourseForm adminNewCourseForm, BindingResult result, Model model) {
+        if (result.hasErrors()){
+            return createForm(model);
+        }
+        Course course = adminNewCourseForm.toModel(instructorRepository, subCategoryRepository, adminNewCourseForm);
+        courseRepository.save(course);
+        return "redirect:/admin/courses/" + course.getCategoryUrlCode() + "/" + course.getSubCategoryUrlCode();
```

Comment:
O povo costuma a usar `String.format` nesses casos de concatenação.

Não precisa mudar, mas veja o que acha:

```suggestion
        return String.format("redirect:/admin/courses/%s/%s,  course.getCategoryUrlCode(),  course.getSubCategoryUrlCode()");
```

Infelizmente, no Java não temos String Templates como outras linguagens. :cry:

Reply:
Bem melhor!